### DISCUSSION TITLE
A new documentation website is in progress, help us improve it

### DISCUSSION THREAD

**Discussion by filips123** [filips123/PWAsForFirefox#335] [General]
After more than a year since originally mentioned in #116, I finally have a work-in-progress documentation website. Once it becomes ready, it will replace the wiki and serve as a main documentation source for this project.

### About the new documentation

The website source is currently available in the [`wip-documentation`](https://github.com/filips123/PWAsForFirefox/tree/wip-documentation/docs) branch in the `docs` directory. It's built with [Material for MkDocs](https://squidfunk.github.io/mkdocs-material/) and will be deployed to GitHub Pages or ReadTheDocs (I haven't decided yet) once ready. The reason for using the website over existing wiki is because this will allow better search, additional Markdown features, and enable PR for improvements from the community.

Unfortunately, I won't have much time for around two months, so I can't promise when it will be ready. Most things are finished, but there are still some things that I would like to add. You can search for "TODO" to find them.

### How can you help

You can preview the documentation yourself by cloning the [`wip-documentation`](https://github.com/filips123/PWAsForFirefox/tree/wip-documentation/docs) branch, installing dependencies (`pip install -r requirements.txt`) and building the website (`mkdocs serve`).

Please check the documentation, share what you think about it and what should still be improved. Do you like the new documentation structure? What do you think is still missing?

Please use this discussion thread for your suggestions and other discussions. You can also submit PRs to the correct branch for direct improvements.

### Notes

Things to mention in frequently asked questions or troubleshooting guides:

* If chrome modifications are not loaded, temporarily enable "always patch runtime and profile". This might happen if profile patches are applied but runtime patches aren't, for example after runtime reinstallation. I will also check if there is any good way to fix this.

* When creating a new profile template, make sure the source browser is closed before copying the directory to a template. This is related to the next thing on this list.

* If the profile template cannot be copied, make sure it has the correct permissions and remove any lock files (#337). I don't think I can remove those files automatically if they can't even be copied, so users have to remove them manually. However, this should not happen often.

* Explain problems related to sandboxed browsers in more detail. For example, https://github.com/filips123/PWAsForFirefox/issues/325#issuecomment-1564131231.

**Comment by filips123** [comment]
The new website is now available at https://pwasforfirefox.filips.si/. Please check it out and share any feedback.

**Reply by black7375** [reply]
Awesome. The site looks neat.

**Comment by j03ll0b0** [comment]
I've built Mercury on mac. It's a fork of Firefox with several enhancements https://github.com/Alex313031/Mercury/releases and I want to use it for handling PWA installations, I followed the instructions but the extension keeps wanting Firefox.app instead. Sorry if I'm writing in the wrong place.

**Reply by filips123** [reply]
Do you want to use Mercury as your main/default browser, or as browser where PWAs will be running?

**Comment by j03ll0b0** [comment]
The only thing keeping me from using it as my main browser is that I have my PWA’s with Thorium and I can not use Mercury to create my PWA’s, I tried and ended up with Firefox.app inside the folder where it downloads the copy for the PWA’s, I tried changing that around to make it use the copy of Mercury I have in my Apps system folder but then it made the wizard go back to the step where you have to make the setup for it, so I assume it didn’t like it or it checks on the version and does as if there was no usable version. I tried putting Mercury.app there, I tried doing the ln -sf, I tried putting the executable, nothing worked. So I guess it doesn’t likes Mercury and also I’m using tip-of-tree v.124, I build myself with the help of Alex313031. You should give his great project a chance to be able to use your great extension.

**Reply by filips123** [reply]
These are a few things you need to make sure are set up correctly:

* The runtime should be installed in `$HOME/Library/Application Support/firefoxpwa/runtime/`.
* The app bundle should be named `Firefox.app`, so `$HOME/Library/Application Support/firefoxpwa/runtime/Firefox.app/`.
* Inside the app bundle, there should be executable file `Contents/MacOS/firefox` (you should symlink it to the Mercury executable if it's different).
* Inside the app bundle, there should be `Contents/Resources/application.ini` INI file. If it doesn't exist, you can put something like this inside it:
   ```ini
   [App]
   Vendor=Mozilla
   Name=Firefox
   RemotingName=firefox
   Version=121.0
   ```

This is what the program checks to see if the runtime is installed. If you already have everything set up according to this, something else might be the problem, but I don't know what exactly.

**Comment by j03ll0b0** [comment]
I appreciate your time and consideration. You’re being very kind. Got a couple questions:
Runtime , is it just the Mercury.app, symlinked?
App bundle, are we talking about two different things? Isn’t the same as the runtime? Also symlinked to the Mercury.app I’ve got in /Applications?

The information is very insightful, I’m very exited about being able to get Mercury PWA’s working!!!

**Comment by j03ll0b0** [comment]
It’s running! But I get the following error at trying to install zoho workplace.

Thanks for your hard work!
￼

**Reply by filips123** [reply]
> Runtime , is it just the Mercury.app, symlinked?

I think symlinks don't work correctly, so it's probably better if you just copy it.

> App bundle, are we talking about two different things? Isn’t the same as the runtime?

The app bundle is specifically the `Mercury.app` directory (that appears as a `Mercury` app on macOS). So, this directory/bundle/app should be named `Firefox.app` and copied into `$HOME/Library/Application Support/firefoxpwa/runtime/`.

The directory structure should then be something like:

```
/Users/user/Library/Application Support/firefoxpwa/runtime        <--- The base runtime directory
/Users/user/Library/Application Support/firefoxpwa/runtime/Firefox.app        <--- The app bundle that you copy and rename from Mercury
...
/Users/user/Library/Application Support/firefoxpwa/runtime/Firefox.app/Contents        <--- Other stuff inside the app bundle
...
/Users/user/Library/Application Support/firefoxpwa/runtime/Firefox.app/Contents/MacOS/firefox        <--- Make sure it's named firefox
/Users/user/Library/Application Support/firefoxpwa/runtime/Firefox.app/Contents/Resources/application.ini        <-- Make sure this one exists and has a valid version key
...
```

>  But I get the following error at trying to install zoho workplace.

Did you try to attach some screenshot? Because unfortunetely it seems it didn't work. You will probably need to attach it directly from GitHub website.

**Comment by j03ll0b0** [comment]
It’s working!!!

￼
Thank you. Great project.

I’ll let Alex313031 know about the wonders of using Mercury with PWAs.

**Comment by j03ll0b0** [comment]
There is a rather funny thing going on, now, instead of Mercury on the menu bar, it says WhatsApp, or Zoho Mail.

**Reply by filips123** [reply]
Does it say that in the PWA or in the main browser? If it's in PWA, it's intentional, but if it's in the main browser, it might be because you were using symlinks, so the app name was overwritten.

**Comment by j03ll0b0** [comment]
Ok, oh, yes, when it’s the app is great to have its name changed to the intended app I’m on, I know that, but whenever I’m on the main app, it reads the last I used I guess? So it might be what you said, since I’m using symlinks as suggested in your extension, “ln -sf…” so… the only way around is to put a full app and modify it as described, correct?

**Reply by filips123** [reply]
Yeah, on macOS symlinks won't work properly, because PWAsForFirefox changes the app name (and [a few other properties](https://github.com/filips123/PWAsForFirefox/blob/main/native/src/components/runtime.rs#L313-L362)) inside `Info.plist` each time you launch a web app, so these changes are then also applied to the main browser.

You could probably get around this by symlinking just specific files and directories that are not affected by this, but unfortunately, I don't know which exactly are these files, so it's probably the easiest if you just fully copy the runtime. In the future, there might be a more automated way to do these symlinks, but I'm not sure yet.

**Comment by j03ll0b0** [comment]
I made a video and although it was only 17mb it didn’t allow to go saying it was too long, well… I’ll just leave it as it is and when I compile it again I will then make the change and put a full copy in there because since I compiled it myself I don’t have the DMG anymore, I cleaned up everything, no biggie, I love your extension, I use it every day now and won’t go back to Thorium anymore since I like my copy of Mercury compiled with LTO so much, it is one of the best mods I’have ever used, and that along your extension gave me limitless possibilities. Hope I can collaborate more than just words in the future though. I’m trying now to build with also PGO but it seems almost imposible just yet.

> Begin forwarded message:
> 
> From: Néstor Espinoza ***@***.***>
> Subject: Re: [filips123/PWAsForFirefox] A new documentation website is in progress, help us improve it (Discussion #335)
> Date: February 4, 2024 at 11:37:59 AM GMT-4
> To: filips123/PWAsForFirefox ***@***.***>
> 
> 
> Yes ! It’s exactly what you’re saying. It’s not annoying since the functionality doesn’t seem to be affected, it’s only cosmetic. Nevertheless, it would be nice to be able to have it actually display correctly, if not possible and the only solution is to not use shortcusts and instead put a full copy of Mercury in the resources folder then I think I will just leave it like that and just laugh at it really loud. Hehe...
>

**Comment by cakira** [comment]
Hello! Thank you for the PWA for Firefox extension. I really appreciate the time and effort you’ve put into developing it.
I came across and installed the extension today, and I was able to install the web app I wanted (🎉). It's already working nicely for what I need.

That said, I’m still figuring out some of the configurations. The Extension Guide mentions that _'All profiles can be accessed and managed through the "profiles" tab in the extension'_, but I haven’t been able to find this tab. I checked in Firefox’s hamburger menu -> Add-ons and themes -> Progressive Web Apps for Firefox -> Three-dot menu. For some extensions, I see options like "Remove, Options, Report, Manage," but for Progressive Web Apps, I only see "Remove, Report, Manage." Could you guide me on where to find these tabs?

Also, just a bit of feedback: it took me a moment to understand the concept of the _"app browser (the custom Firefox installation that runs installed web apps)."_ I only grasped it after opening the web app and noticing that Firefox's menus were in English, even though my default language is Portuguese. Others might face the same initial confusion.

Overall, the extension is working great for me, and I’m really impressed with how customizable it is. I just wanted to share my experience in case it helps others. Thanks again for creating such an awesome tool!

**Reply by filips123** [reply]
> The Extension Guide mentions that 'All profiles can be accessed and managed through the "profiles" tab in the extension', but I haven’t been able to find this tab.

I mean the extension popup that can be accessed from the toolbar widget:

![slika](https://github.com/user-attachments/assets/eb258f21-0512-4d98-9861-0d9075c6fd6e)

> Also, just a bit of feedback: it took me a moment to understand the concept of the "app browser (the custom Firefox installation that runs installed web apps)." I only grasped it after opening the web app and noticing that Firefox's menus were in English, even though my default language is Portuguese. Others might face the same initial confusion.

Do you have any suggestions or ideas how to make this more clear?

**Reply by cakira** [reply]
Wow!
Thank you very much! I've been using extensions for several years, and I never realized that the extension names were clickable. 🤯

As for ideas... I have a rough concept, so I might be off the mark here, but it seems that, unfortunately, most people (myself included) don't read texts longer than 280 characters, let alone instruction manuals. So, one approach could be to create a very short and visual "quick guide" with just one example to get started (maybe a video). Further information could be available in a "reference manual."

I can't promise, but I'll try to go through the entire installation and usage process tonight or tomorrow night to gather more ideas and make sure I'm not speaking nonsense.

**Comment by rtfmoz2** [comment]
I've provided an installation guide for windows users in #721 and created the start of the install guide. Will need to flesh out the details for other platforms. I think I need a small diagram that describes how it works with firefox. To give the end user a general understanding of dependencies of the major components.

Firefox -> PWAForForefox Extension <-> PWAForFirefox application -> Customised Firefox -> PWA

Terms:  Used application instead of native. Native is just an application type and people understand "application" already.
